File: app/src/test/java/com/vrem/wifianalyzer/wifi/timegraph/TimeGraphCacheTest.java
/*
 * WiFiAnalyzer
 * Copyright (C) 2019  VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

package com.vrem.wifianalyzer.wifi.timegraph;

import com.vrem.wifianalyzer.wifi.band.WiFiWidth;
import com.vrem.wifianalyzer.wifi.graphutils.GraphConstants;
import com.vrem.wifianalyzer.wifi.model.WiFiDetail;
import com.vrem.wifianalyzer.wifi.model.WiFiSignal;

import org.apache.commons.collections4.Closure;
import org.apache.commons.collections4.IterableUtils;
import org.apache.commons.lang3.StringUtils;
import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class TimeGraphCacheTest {

    private TimeGraphCache fixture;

    @Before
    public void setUp() {
        fixture = new TimeGraphCache();
    }

    @Test
    public void testAll() {
        // setup
        List<WiFiDetail> expected = withWiFiDetails();
        // execute
        Set<WiFiDetail> actual = fixture.getWiFiDetails();
        // validate
        assertEquals(expected.size(), actual.size());
    }

    @Test
    public void testActive() {
        // setup
        List<WiFiDetail> expected = withWiFiDetails();
        // execute
        Set<WiFiDetail> actual = fixture.active();
        // validate
        assertEquals(expected.size() - 1, actual.size());
        assertFalse(actual.contains(expected.get(0)));
    }

    @Test
    public void testClear() {
        // setup
        List<WiFiDetail> expected = withWiFiDetails();
        // execute
        fixture.clear();
        // validate
        Set<WiFiDetail> actual = fixture.getWiFiDetails();
        assertEquals(expected.size() - 1, actual.size());
        assertFalse(actual.contains(expected.get(0)));
    }

    @Test
    public void testReset() {
        // setup
        List<WiFiDetail> expected = withWiFiDetails();
        // execute
        fixture.reset(expected.get(0));
        // validate
        Set<WiFiDetail> actual = fixture.getWiFiDetails();
        assertEquals(expected.size(), actual.size());
        assertTrue(actual.contains(expected.get(0)));
    }

    private WiFiDetail withWiFiDetail(String SSID) {
        return new WiFiDetail(SSID, "BSSID", StringUtils.EMPTY,
            new WiFiSignal(100, 100, WiFiWidth.MHZ_20, 5, true));
    }

    private List<WiFiDetail> withWiFiDetails() {
        List<WiFiDetail> results = new ArrayList<>();
        for (int i = 0; i < 4; i++) {
            WiFiDetail wiFiDetail = withWiFiDetail("SSID" + i);
            results.add(wiFiDetail);
        }
        IterableUtils.forEach(results, new AddClosure());
        for (int i = 0; i < GraphConstants.MAX_NOTSEEN_COUNT; i++) {
            fixture.add(results.get(0));
        }
        return results;
    }

    private class AddClosure implements Closure<WiFiDetail> {
        @Override
        public void execute(WiFiDetail wiFiDetail) {
            fixture.add(wiFiDetail);
        }
    }
}

File: app/src/test/kotlin/com/vrem/wifianalyzer/wifi/timegraph/TimeGraphCacheTest.kt
/*
 * WiFiAnalyzer
 * Copyright (C) 2015 - 2025 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
package com.vrem.wifianalyzer.wifi.timegraph

import com.vrem.wifianalyzer.wifi.graphutils.MAX_NOT_SEEN_COUNT
import com.vrem.wifianalyzer.wifi.model.*
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class TimeGraphCacheTest {
    private val fixture = TimeGraphCache()

    @Test
    fun all() {
        // setup
        val expected = withWiFiDetails()
        // execute
        val actual = fixture.wiFiDetails
        // validate
        assertThat(actual).hasSize(expected.size)
    }

    @Test
    fun active() {
        // setup
        val expected = withWiFiDetails()
        // execute
        val actual = fixture.active()
        // validate
        assertThat(actual).hasSize(expected.size - 1)
        assertThat(actual).doesNotContain(expected[0])
    }

    @Test
    fun clear() {
        // setup
        val expected = withWiFiDetails()
        // execute
        fixture.clear()
        // validate
        val actual = fixture.wiFiDetails
        assertThat(actual).hasSize(expected.size - 1)
        assertThat(actual).doesNotContain(expected[0])
    }

    @Test
    fun reset() {
        // setup
        val expected = withWiFiDetails()
        // execute
        fixture.reset(expected[0])
        // validate
        val actual = fixture.wiFiDetails
        assertThat(actual).hasSize(expected.size)
        assertThat(actual).contains(expected[0])
    }

    private fun withWiFiDetail(ssid: String): WiFiDetail {
        return WiFiDetail(
            WiFiIdentifier(ssid, "BSSID"),
            WiFiSecurity.EMPTY,
            WiFiSignal(100, 100, WiFiWidth.MHZ_20, 5)
        )
    }

    private fun withWiFiDetails(): List<WiFiDetail> {
        val results: MutableList<WiFiDetail> = mutableListOf()
        for (i in 0..3) {
            val wiFiDetail = withWiFiDetail("SSID$i")
            results.add(wiFiDetail)
        }
        results.forEach { fixture.add(it) }
        for (i in 0 until MAX_NOT_SEEN_COUNT) {
            fixture.add(results[0])
        }
        return results
    }

}